packages/actor-rdf-join-wrap-stream/lib/ActorRdfJoinWrapStream.ts

```diff
@@ -0,0 +1,90 @@
+import type { MediatorIteratorTransform } from '@comunica/bus-iterator-transform';
+import type {
+  IActionRdfJoin,
+  IActorRdfJoinOutputInner,
+  IActorRdfJoinArgs,
+  MediatorRdfJoin }
+  from '@comunica/bus-rdf-join';
+import { ActorRdfJoin, KEY_CONTEXT_WRAPPED_RDF_JOIN } from '@comunica/bus-rdf-join';
+import type { IMediatorTypeJoinCoefficients } from '@comunica/mediatortype-join-coefficients';
+import type { IQueryOperationResultBindings, MetadataBindings } from '@comunica/types';
+import type * as RDF from '@rdfjs/types';
+import type { AsyncIterator } from 'asynciterator';
+
+/**
+ * A comunica Wrap Stream RDF Join Actor.
+ */
+export class ActorRdfJoinWrapStream extends ActorRdfJoin {
+  public readonly mediatorJoin: MediatorRdfJoin;
+  public readonly mediatorIteratorTransform: MediatorIteratorTransform;
+
+  public constructor(args: IActorRdfJoinWrapStreamArgs) {
+    super(args, {
+      logicalType: 'inner',
+      physicalName: 'wrap-stream',
+      limitEntries: 0,
+      limitEntriesMin: true,
+      canHandleUndefs: true,
+      isLeaf: false,
+    });
+  }
+
+  public override async test(action: IActionRdfJoin): Promise<IMediatorTypeJoinCoefficients> {
+    if (action.context.get(KEY_CONTEXT_WRAPPED_RDF_JOIN)) {
+      throw new Error('Unable to wrap join operation multiple times');
+    }
+
+    const metadatas = await ActorRdfJoin.getMetadatas(action.entries);
+    return this.getJoinCoefficients(action, metadatas);
+  }
+
+  public override async getOutput(action: IActionRdfJoin): Promise<IActorRdfJoinOutputInner> {
+    // Prevent infinite recursion. In consequent query operation calls this key is set to false
+    // To allow the operation to wrap ALL rdf-join runs
+    action.context = ActorRdfJoin.setContextWrapped(action.context, true);
+    const result: IQueryOperationResultBindings = await this.mediatorJoin.mediate(action);
+
+    const { stream, streamMetadata } = (await this.mediatorIteratorTransform.mediate(
+      {
+        operation: action.type,
+        stream: result.bindingsStream,
+        streamMetadata: result.metadata,
```

Comment:
Why both metadata and streamMetadata?

Reply:
The intention is to allow users to change the `streamMetadata` (the metadata function given by the join result) in iterator transform actors. The `metadata` field is intended as general metadata including more than just the stream metadata. If that makes sense.

Reply:
I may be misunderstanding something, but if users want to change the metadata, this could be done in the return value of the mediator, right?
I suspect we may only need `streamMetadata`.

Reply:
I think the intention was to allow actors on `bus-transform-iterator` to apply a transformation on the intermediate iterator's `metadata` function. For example, if you filter some results and want to update the estimated cardinality. In this case `streamMetadata` is the `metadata` function of the intermediate iterator (which may be somewhat confusingly named).

Reply:
Ok, but then `streamMetadata` is all we need, and `metadata` can be removed (and perhaps then `streamMetadata` can be renamed to `metadata`)?

Reply:
But some metadata could be useful to pass through to the `iterator-transform` actors that are not in `streamMetadata`, like in `rdf-join-wrap-stream`, the number of entries in the wrapped action shows how many joins were already performed. Where would you suggest this be added?

Reply:
Aha, now I understand you.
I didn't understand the need for two types of metadata fields, but now I do.

Instead of passing custom fields such as the number of join entries in `metadata`, why not omit this custom metadata, and instead directly pass the original `action` via e.g. `originalAction`, to allow the user to derive the number of join entries like that?

Reply:
Right, that works!